modules/dnn/src/darknet/darknet_io.cpp

```diff
@@ -0,0 +1,638 @@
+            std::string layer_type = layer_params["type"];
+
+            if (layer_type == "convolutional")
+            {
+                int kernel_size = getParam<int>(layer_params, "size", -1);
+                int pad = getParam<int>(layer_params, "pad", 0);
+                int stride = getParam<int>(layer_params, "stride", 1);
+                int filters = getParam<int>(layer_params, "filters", -1);
+                std::string activation = getParam<std::string>(layer_params, "activation", "linear");
+                bool batch_normalize = getParam<int>(layer_params, "batch_normalize", 0) == 1;
+                if(activation != "linear" && activation != "leaky")
+                    CV_Error(cv::Error::StsParseError, "Unsupported activation: " + activation);
+
+                // correct the strange value of pad=1 for kernel_size=1 in the Darknet cfg-file
+                if (kernel_size < 3) pad = 0;
+
+                CV_Assert(kernel_size > 0 && filters > 0);
+                CV_Assert(current_shape.input_channels > 0);
+
+                setParams.setConvolution(kernel_size, kernel_size, pad, pad, stride, stride, filters, current_shape.input_channels,
+                    batch_normalize, activation == "leaky");
+
+                current_shape.input_channels = filters;
+                current_shape.input_h = (current_shape.input_h + 2 * pad - kernel_size) / stride + 1;
+                current_shape.input_w = (current_shape.input_w + 2 * pad - kernel_size) / stride + 1;
+            }
+            else if (layer_type == "maxpool")
+            {
+                int kernel_size = getParam<int>(layer_params, "size", 2);
+                int stride = getParam<int>(layer_params, "stride", 2);
+                int pad = getParam<int>(layer_params, "pad", 0);
+                setParams.setMaxpool(kernel_size, pad, stride, current_shape.input_channels);
+
+                current_shape.input_w = current_shape.input_w / stride;
+                current_shape.input_h = current_shape.input_h / stride;
+            }
+            else if (layer_type == "route")
+            {
+                std::string bottom_layers = getParam<std::string>(layer_params, "layers", "");
+                CV_Assert(!bottom_layers.empty());
+                std::vector<int> layers_vec = getNumbers<int>(bottom_layers);
+
+                current_shape.input_channels = 0;
+                for (size_t k = 0; k < layers_vec.size(); ++k) {
+                    layers_vec[k] += layers_counter;
+                    current_shape.input_channels += net->out_shape_vec[layers_vec[k]].input_channels;
+                }
+                current_shape.input_w = net->out_shape_vec[layers_vec[0]].input_w;
+                current_shape.input_h = net->out_shape_vec[layers_vec[0]].input_h;
+
+                setParams.setConcat(layers_vec.size(), layers_vec.data());
+            }
+            else if (layer_type == "reorg")
```

Comment:
I'm a bit confused about reorg layer. Let the input is:
```
channel_0  channel_1  channel_2  channel_3
0 1        4 5        8 9        c d
2 3        6 7        a b        e f
```
and reorgStride = 2. So an output shape is 4x4x1 and values are:
```
output
1 4 1 5
8 c 9 d
2 6 3 7
a e b f
```
?

Reply:
I left unchanged a bit strange original implementation of this layer. It increases the field of view of each final activation.
Reshape: **26 x 26 x 64 -> 13 x 13 x 256**
![reorg](https://user-images.githubusercontent.com/4096485/30909091-8c787fe8-a388-11e7-958f-f348ec1a524d.jpg)

```
input
    0, 1, 2, 3,
    4, 5, 6, 7, 
    8, 9, a, b,
    c, d, e, f
```

```
output
channel_0  channel_1  channel_2  channel_3
0 2        1 3        4 6        5 7
8 a        9 b        c e        d f
```
* OpenCV C++ example: http://coliru.stacked-crooked.com/a/eb13942be083fa3d
* Darknet C++ example: http://coliru.stacked-crooked.com/a/225d7fb1f25b286c

----
* Param `reverse` usually absent in the cfg-file of model, so `reverse = 0` by default https://github.com/pjreddie/darknet/blob/master/src/parser.c#L387

* Then `l.reverse = reverse;` https://github.com/pjreddie/darknet/blob/master/src/reorg_layer.c#L28

* So the function `reorg()` is called with `forward=0`: https://github.com/pjreddie/darknet/blob/8215a8864d4ad07e058acafd75b2c6ff6600b9e8/src/reorg_layer.c#L108

* `reorg_cpu()`-implementation uses `out[in_index] = x[out_index];`: https://github.com/pjreddie/darknet/blob/master/src/blas.c#L25

Reply:
Thanks! Anyway I suggest to replace it as a single layer or think how we can do the same transformations using existing ones (Permute, Reshape). Reshape layer doesn't change the data by definition neither in one of the frameworks.

Reply:
@dkurt I added reorg as separate layer `reorg_layer.cpp`